File: 316_Remove_Duplicate_Letters/316_1.cpp
class Solution {
public:
    string removeDuplicateLetters(string s) {
        stack<char> stack;
        int count[255];
        memset(count, 0, sizeof(count));
        for (char c : s) {
            count[c]++;
        }
        unordered_set<char> seen;
        for (char c : s) {
            if (seen.find(c) == seen.end()) {
                while (!stack.empty() && c < stack.top() && count[stack.top()] > 0) {
                    seen.erase(stack.top());
                    stack.pop();
                }
                stack.emplace(c);
                seen.insert(c);
            }
            count[c]--;
        }

        string res;
        while (!stack.empty()) {
            res.push_back(stack.top());
            stack.pop();
        }
        reverse(res.begin(), res.end());
        return res;
    }
};

File: 316_Remove_Duplicate_Letters/316_1.py
from collections import Counter

class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        stack = []
        count = Counter(s)
        seen = set()
        for c in s:
            if c not in seen:
                while stack and c < stack[-1]  and count[stack[-1]] > 0:
                    seen.remove(stack.pop())
                seen.add(c)
                stack.append(c)
            count[c] -= 1
        return "".join(stack)